resources/js/features/users-edit/UserEdit.form.tsx

```diff
@@ -42,91 +50,121 @@ export namespace UserEditForm {
 
   interface Props extends FormParams {
     onSuccess?: () => void;
+    onCancel?: () => void;
   }
 
   export const Component: Component<Props> = (props) => {
     const t = useLangFunc();
+    const statusQuery = createQuery(() => User.statusQueryOptions);
     const userQuery = createQuery(() => Admin.userQueryOptions(props.userId));
     const user = () => userQuery.data;
     const invalidate = Admin.useInvalidator();
-    const mutation = createMutation(() => ({
-      mutationFn: Admin.updateUser,
-      onSuccess() {
-        invalidate.users();
-        props.onSuccess?.();
-      },
-    }));
-    async function onSubmit(values: Output) {
-      await mutation.mutateAsync({
-        id: user()!.id,
+    const userMutation = createMutation(() => ({mutationFn: Admin.updateUser}));
+    const membersUpdater = UserMembersEdit.useMembersUpdater();
+
+    async function updateUser(values: Output, ctx: SubmitContext<Output>) {
+      const oldUser = user()!;
+      if (oldUser.id === statusQuery.data?.user.id && oldUser.hasGlobalAdmin && !values.hasGlobalAdmin) {
+        ctx.setErrors("hasGlobalAdmin", t("forms.user_edit.validation.cannot_remove_own_global_admin"));
+        return;
+      }
+      // First mutate the user fields (without the members).
+      await userMutation.mutateAsync({
+        id: oldUser.id,
         name: values.name,
         ...(values.email
           ? {
               email: values.email,
               hasEmailVerified: values.hasEmailVerified,
               ...(values.hasPassword
-                ? user()?.hasPassword && !values.password
+                ? oldUser.hasPassword && !values.password
                   ? // The user had password already and it is not changed.
                     {}
                   : // New password or a password change.
                     {password: values.password, passwordExpireAt: null}
                 : {password: null}),
             }
           : {email: null}),
+        hasGlobalAdmin: values.hasGlobalAdmin,
       });
+      // If the user mutation succeeded, await all the members mutations.
+      try {
+        await Promise.all(membersUpdater.getUpdatePromises(oldUser, values.members));
```

Comment:
Lepiej użyć Promise.allSetttled, to w miarę świeża funkcja.

Cytując z MDN (bo po ang. łatwiej):

> Promise.all() will reject immediately upon any of the input promises rejecting. In comparison, the promise returned by [Promise.allSettled()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/allSettled) will wait for all input promises to complete, regardless of whether or not one rejects. Use allSettled() if you need the final result of every promise in the input iterable.

źródło: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/all

Tu nie potrzebujemy odpowiedzi z requestów, ale warto poczekać z `invalidate.users();` do momentu, aż wszystkie zapytania zostaną zrealizowane (obojętnie z jakim skutkiem).

Obecnie, mając tu Promise.all, `invalidate.users();` wykona się także w momencie, gdy pierwszy z requestów zakończy się niepowodzeniem a reszta będzie jeszcze w toku, czyli inwalidacja mogłaby nastąpić przed aktualizacją danych na serwerze (a to jest niepożądane, po to inwalidujemy żeby mieć aktualne dane).

Reply:
No słusznie, użyłem all bo nie pomyślałem o tym przypadku. Done.